lib/internal/modules/esm/loader.js

```diff
@@ -194,34 +237,45 @@ class ESMLoader {
   /**
    * Collect custom/user-defined hook(s). After all hooks have been collected,
    * calls global preload hook(s).
-   * @param {object | object[]} customLoaders A list of exports from
-   * user-defined loaders (as returned by ESMLoader.import()).
+   * @param {KeyedExports} customLoaders
+   *  A list of exports from user-defined loaders (as returned by
+   *  ESMLoader.import()).
    */
   async addCustomLoaders(
     customLoaders = [],
   ) {
-    if (!ArrayIsArray(customLoaders)) customLoaders = [customLoaders];
-
     for (let i = 0; i < customLoaders.length; i++) {
-      const exports = customLoaders[i];
+      const {
+        exports,
+        url,
+      } = customLoaders[i];
       const {
         globalPreloader,
         resolver,
         loader,
       } = ESMLoader.pluckHooks(exports);
 
-      if (globalPreloader) ArrayPrototypePush(
-        this.#globalPreloaders,
-        FunctionPrototypeBind(globalPreloader, null), // [1]
-      );
-      if (resolver) ArrayPrototypePush(
-        this.#resolvers,
-        FunctionPrototypeBind(resolver, null), // [1]
-      );
-      if (loader) ArrayPrototypePush(
-        this.#loaders,
-        FunctionPrototypeBind(loader, null), // [1]
-      );
+      if (globalPreloader) {
+        const hook = ObjectCreate(null);
+        hook.url = url;
+        hook.fn = FunctionPrototypeBind(globalPreloader); // [1]
+
+        this.#globalPreloaders.push(hook);
```

Comment:
We should be using primordials here and below
```suggestion
        ArrayPrototypePush(
          this.#globalPreloaders,
          {
            url,
            fn: FunctionPrototypeBind(globalPreloader), // [1]
          }
        );
```

Reply:
🤦‍♂️ yes